### PR TITLE
Adding Options for Request Id Policy.

### PR DESCRIPTION
**Pull request by hemanttanwar**
Allow customers set the value in RequestId header with following headers names
1. "x-ms-client-request-id"
2. "x-ms-correlation-request-id"
3. "client-request-id"

Should use something like RequestIDOptions POJO to collect information from user. Reference PR https://github.com/Azure/azure-sdk-for-java/pull/6196/files

Coordinate with PR, since it is waiting for this change. : #6214
Related Issue : #6171

Additional context

This change is inline with overall strategy discussed here #6217

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java [hunk 1]
```diff
@@ -3,24 +3,56 @@
...
  * the unique identifier for the request.
+ * @see RequestIdPolicyOptions
  */
 public class RequestIdPolicy implements HttpPipelinePolicy {
     private static final String REQUEST_ID_HEADER = "x-ms-client-request-id";
 
+    private final RequestIdPolicyOptions requestIdPolicyOptions;
+
+    /**
+     * Creates {@link RequestIdPolicy} with default {@link ExponentialBackoff} as {@link RetryStrategy}and use
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: Update javadoc as this is not related to retry policy.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java [hunk 2]
```diff
@@ -3,24 +3,56 @@
...
+
+    /**
+     * Creates a {@link RequestIdPolicy} with the provided {@link RequestIdPolicyOptions}.
+     * User can specify {@link java.util.function.Function} to dynamically generated id for various id headers.
+     *
+     * @param requestIdPolicyOptions with given {@link RetryPolicyOptions}.
+     * @throws NullPointerException if {@code RequestIdPolicyOptions} is {@code null}.
+     */
+    public RequestIdPolicy(RequestIdPolicyOptions requestIdPolicyOptions) {
+        this.requestIdPolicyOptions = Objects.requireNonNull(requestIdPolicyOptions,
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: Null check is not necessary as the no-args constructor sets the `requestIdPolicyOptions` to null anyway.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Immutable Configuration options for {@link RequestIdPolicy}.
+ * User can provide a {@link Supplier} to generated id for various id headers to be used in
+ * {@link com.azure.core.http.HttpRequest}. Examples of such headers are 'x-ms-client-request-id
+ * or 'x-ms-correlation-request-id' etc.
+ */
+@Immutable
+public class RequestIdPolicyOptions {
+
+    private volatile Supplier<HttpHeaders> idHeaderSupplier;
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java]: Why `volatile`?

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java [hunk 4]
```diff
@@ -3,24 +3,56 @@
...
+    }
+
+
     @Override
     public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
-        String requestId = context.getHttpRequest().getHeaders().getValue(REQUEST_ID_HEADER);
-        if (requestId == null) {
+        if (!Objects.isNull(requestIdPolicyOptions)) {
+            HttpHeaders httpHeaders =  requestIdPolicyOptions.getIdHeaderSupplier().get();
+            if (!Objects.isNull(httpHeaders)) {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: A random header id should be added if `httpHeaders` is null.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+     * @param idHeaderSupplier {@link Supplier} to provide dynamically generated id for various request id headers.
+     * Examples of such headers are 'x-ms-client-request-id or 'x-ms-correlation-request-id' etc.
+     */
+    public RequestIdPolicyOptions(Supplier<HttpHeaders> idHeaderSupplier) {
+        this.idHeaderSupplier = Objects.requireNonNull(idHeaderSupplier,
+            "'idHeaderSupplier' cannot be null.");
+    }
+
+    /**
+     *
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java]: Missing javadoc

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    private volatile Supplier<HttpHeaders> idHeaderSupplier;
+
+    /**
+     * Creates {@link RequestIdPolicyOptions} with provided {@code idHeaderSupplier }.
+     *
+     * @param idHeaderSupplier {@link Supplier} to provide dynamically generated id for various request id headers.
+     * Examples of such headers are 'x-ms-client-request-id or 'x-ms-correlation-request-id' etc.
+     */
+    public RequestIdPolicyOptions(Supplier<HttpHeaders> idHeaderSupplier) {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java]: name suggestion: `requestIdSupplier`

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.function.Supplier;
+
+/**
+ * Immutable Configuration options for {@link RequestIdPolicy}.
+ * User can provide a {@link Supplier} to generated id for various id headers to be used in
+ * {@link com.azure.core.http.HttpRequest}. Examples of such headers are 'x-ms-client-request-id
+ * or 'x-ms-correlation-request-id' etc.
+ */
+@Immutable
+public class RequestIdPolicyOptions {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java]: Do we need a separate type for this? Can the `Supplier` instead be passed to `RequestIdPolicy` constructor directly?

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java [hunk 8]
```diff
@@ -3,24 +3,56 @@
...
 import com.azure.core.http.HttpPipelineNextPolicy;
 import com.azure.core.http.HttpResponse;
 import reactor.core.publisher.Mono;
 
+import java.util.Objects;
 import java.util.UUID;
 
 /**
- * The pipeline policy that puts a UUID in the request header. Azure uses the request id as
+ * The pipeline policy that puts a UUID or user provided id in the request header. Azure uses the request id as
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: With this change, we can't guarantee that the request-id is a unique identifier. User-provided request ids may not be unique.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java [hunk 9]
```diff
@@ -3,24 +3,56 @@
...
+     */
+    public RequestIdPolicy() {
+        this.requestIdPolicyOptions = null;
+    }
+
+    /**
+     * Creates a {@link RequestIdPolicy} with the provided {@link RequestIdPolicyOptions}.
+     * User can specify {@link java.util.function.Function} to dynamically generated id for various id headers.
+     *
+     * @param requestIdPolicyOptions with given {@link RetryPolicyOptions}.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: Fix javadoc - `RequestIdPolicyOptions`

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java [hunk 10]
```diff
@@ -3,24 +3,56 @@
...
+     * Creates {@link RequestIdPolicy} with default {@link ExponentialBackoff} as {@link RetryStrategy}and use
+     * 'retry-after-ms' in {@link HttpResponse} header for calculating retry delay.
+     */
+    public RequestIdPolicy() {
+        this.requestIdPolicyOptions = null;
+    }
+
+    /**
+     * Creates a {@link RequestIdPolicy} with the provided {@link RequestIdPolicyOptions}.
+     * User can specify {@link java.util.function.Function} to dynamically generated id for various id headers.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: This should be `RequestIdPolicyOptions` instead of `Function`. Also, import packages and use class names instead of fully qualified names.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java [hunk 11]
```diff
@@ -3,24 +3,56 @@
...
-        if (requestId == null) {
+        if (!Objects.isNull(requestIdPolicyOptions)) {
+            HttpHeaders httpHeaders =  requestIdPolicyOptions.getIdHeaderSupplier().get();
+            if (!Objects.isNull(httpHeaders)) {
+                httpHeaders.stream().forEach(httpHeader ->
+                    context.getHttpRequest().getHeaders().put(httpHeader.getName(), httpHeader.getValue())
+                );
+            }
+        } else {
             context.getHttpRequest().getHeaders().put(REQUEST_ID_HEADER, UUID.randomUUID().toString());
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: This will now overwrite the requestId that was already in `context`. Existing requestId should be preserved. Random id should be used only when the requestId in context is null.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java [hunk 12]
```diff
@@ -3,24 +3,56 @@
...
+
     @Override
     public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
-        String requestId = context.getHttpRequest().getHeaders().getValue(REQUEST_ID_HEADER);
-        if (requestId == null) {
+        if (!Objects.isNull(requestIdPolicyOptions)) {
+            HttpHeaders httpHeaders =  requestIdPolicyOptions.getIdHeaderSupplier().get();
+            if (!Objects.isNull(httpHeaders)) {
+                httpHeaders.stream().forEach(httpHeader ->
+                    context.getHttpRequest().getHeaders().put(httpHeader.getName(), httpHeader.getValue())
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: This should check for existing headers with the same name in the context. If it already exists, it should not be overwritten.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java (hunk 7 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java]: Agreed on this comment, I don't think this policy will require additional configuration outside of being able to set Request ID headers.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java (hunk 12 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: I think this is a good place to have a discussion, can take it offline if needed.

@JonathanGiles is there any guidelines around whether a pipeline policy should overwrite or only add if not exists?

Comment (hemanttanwar) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: I am also unsure about guideline on this. I did hear that every new request  ( including retry ) to Azure service should use new new Request ID.  But at the same time it is "optional" in Azure Docs
https://docs.microsoft.com/en-us/rest/api/azure/devops/symbol/requests/create%20requests?view=azure-devops-rest-5.1

Comment (hemanttanwar) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: I just spoke to Jeffery on this  and his comments are "Every retry should have the same ID. This way, we can tell how many retries occurred for a single logical operation."

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java (hunk 7 above)
Comment (hemanttanwar) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicyOptions.java]: agree

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java (hunk 8 above)
Comment (itye-msft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java]: While it's true, it may also be possible that a customer generates their own UUID and may want to correlate it with the ID generated by their system.
I have no idea how the service implements it, but as a user I would expect the service to save the provided ID with the rest of the raw data I sent, so I can analyze it later. Therefore the SDK cannot guarantee uniqueness if user provides the ID, which is fine, if the persistence layer can support it. 
I want to suggest that the code will generate a default ID, and allow the customer to override it based on their need.

**Comment by hemanttanwar** [comment]
Closing this PR this  work was done in a separate PR by adding a constructor `public RequestIdPolicy(String requestIdHeaderName)` . So closing this PR.
